src/pybind/mgr/cephadm/tests/test_scheduling.py

```diff
@@ -625,3 +625,75 @@ def get_hosts_func(label=None, as_hostspec=False):
             get_hosts_func=get_hosts_func,
             get_daemons_func=lambda _: daemons).place()
     assert str(e.value) == expected
+
+class ActiveAssignmentTest(NamedTuple):
+    service_type: str
+    placement: PlacementSpec
+    hosts: List[str]
+    daemons: List[DaemonDescription]
+    expected: List[List[str]]
+
+
+@pytest.mark.parametrize("service_type,placement,hosts,daemons,expected",
+                         [
+                             ActiveAssignmentTest(
+                                 'mgr',
+                                 PlacementSpec(count=2),
+                                 'host1 host2 host3'.split(),
+                                 [
+                                     DaemonDescription('mgr', 'a', 'host1', is_active=True),
+                                     DaemonDescription('mgr', 'b', 'host2'),
+                                     DaemonDescription('mgr', 'c', 'host3'),
+                                 ],
+                                 [['host1', 'host2'], ['host1', 'host3']]
+                             ),
+                             ActiveAssignmentTest(
+                                 'mgr',
+                                 PlacementSpec(count=2),
+                                 'host1 host2 host3'.split(),
+                                 [
+                                     DaemonDescription('mgr', 'a', 'host1'),
+                                     DaemonDescription('mgr', 'b', 'host2'),
+                                     DaemonDescription('mgr', 'c', 'host3', is_active=True),
+                                 ],
+                                 [['host1', 'host3'], ['host2', 'host3']]
+                             ),
+                             ActiveAssignmentTest(
+                                 'mgr',
+                                 PlacementSpec(count=1),
+                                 'host1 host2 host3'.split(),
+                                 [
+                                     DaemonDescription('mgr', 'a', 'host1'),
+                                     DaemonDescription('mgr', 'b', 'host2', is_active=True),
+                                     DaemonDescription('mgr', 'c', 'host3'),
+                                 ],
+                                 [['host2']]
+                             ),
+                             ActiveAssignmentTest(
+                                 'mgr',
+                                 PlacementSpec(count=1),
+                                 'host1 host2 host3'.split(),
+                                 [
+                                     DaemonDescription('mgr', 'a', 'host1'),
+                                     DaemonDescription('mgr', 'b', 'host2'),
+                                     DaemonDescription('mgr', 'c', 'host3', is_active=True),
+                                 ],
+                                 [['host3']]
```

Comment:
what I'm missing is:

* what happens, if there are multiple active daemons
* what happens, if _all_ daemons are active
* what happens, if there are co-located daemons on one host: one active the other one standby + all daemons active
* does explicit placement still work?

Reply:
@sebastian-philipp For clarification, I'm assuming there is only ever 1 active mgr or mds and the rest are standby and my changes to the scheduler assume the same. What would happen if multiple were flagged as active is whichever "active" daemon happens to be first in the HostPlacement object's list of daemons would be considered THE active one and its host would get preference while all other hosts would be treated equally. If you had 4 hosts with mgrs, A, B, C and D, the mgr on hosts A and C were marked active and you tried to scale down to 2 mgrs, either host A or C would be picked first for sure and the second host chosen would be random (you could get any combination besides B and D). If it's meant to be possible to have multiple active mgr or mds I need to adjust this PR to account for that. 

If, on the other hand, you're just concerned with what happens if multiple daemons have accidentally been marked as active somehow, I'll make sure to provide a test for that case.

I don't think co-located daemons should affect the outcome. If one daemon on the host is set as active, regardless of if the other one is not (and assuming no other hosts also have active daemons for reasons mentioned earlier in this comment) that host will get preferred just as it would without a co-located daemon. I'll include a couple tests for this.

+1 on checking explicit placement still works when daemons are marked as active. I'll make sure to include a test for that.

Reply:
> @sebastian-philipp For clarification, 

> I'm assuming there is only ever 1 active mgr or mds and the rest are standby and my changes to the scheduler assume the same. 

MDS actually support multiple active daemons for a given FS. We also have more than active MON at any time.

> What would happen if multiple were flagged as active is whichever "active" daemon happens to be first in the HostPlacement object's list of daemons would be considered THE active one and its host would get preference while all other hosts would be treated equally. If you had 4 hosts with mgrs, A, B, C and D, the mgr on hosts A and C were marked active and you tried to scale down to 2 mgrs, either host A or C would be picked first for sure and the second host chosen would be random (you could get any combination besides B and D). If it's meant to be possible to have multiple active mgr or mds I need to adjust this PR to account for that.
> 
> If, on the other hand, you're just concerned with what happens if multiple daemons have accidentally been marked as active somehow, I'll make sure to provide a test for that case.

:+1: 

> 
> I don't think co-located daemons should affect the outcome.

I know for sure that we (unfortunately) already have to deal with co-located daemons. 

> If one daemon on the host is set as active, regardless of if the other one is not (and assuming no other hosts also have active daemons for reasons mentioned earlier in this comment) that host will get preferred just as it would without a co-located daemon. I'll include a couple tests for this.

:+1: 

> 
> +1 on checking explicit placement still works when daemons are marked as active. I'll make sure to include a test for that.

Reply:
After the most recent update, this should be able to handle multiple hosts with active daemons when placing